File: solutions/javascript/circular-buffer/1/circular-buffer.js
class CircularBuffer {
  constructor(capacity) {
    this.capacity = capacity;
    this.buffer   = Array(capacity);
    this.clear();
  }

  write(item) {
    if (this.buffer.every(i => i != null)) {
      throw new BufferFullError();
    }

    this.buffer[this.writeIndex] = item;
    this.writeIndex = this.advance(this.writeIndex);
  }

  read() {
    if (this.buffer[this.readIndex] == null) {
      throw new BufferEmptyError();
    }

    const value = this.buffer[this.readIndex];
    this.buffer[this.readIndex] = null;
    this.readIndex = this.advance(this.readIndex);

    return value;
  }

  advance = (index) => (index + 1) % this.capacity;

  forceWrite(item) {
    if (this.buffer.some(i => i == null)) {
      this.write(item);
      return;
    }

    this.buffer[this.readIndex] = item;
    this.readIndex = this.advance(this.readIndex);
  }

  clear() {
    this.readIndex  = 0;
    this.writeIndex = 0;
    this.buffer.fill(null);
  }
}

export default CircularBuffer;

export class BufferFullError extends Error {
  constructor() {
    super('Cannot write to a full buffer.');
  }
}

export class BufferEmptyError extends Error {
  constructor() {
    super('Cannot read from an empty buffer.');
  }
}

File: solutions/ruby/circular-buffer/1/circular_buffer.rb
class CircularBuffer
  def initialize(size)
    @buffer      = Array.new(size)
    @read_index  = 0
    @write_index = 0
    @last_index  = size - 1
  end

  def clear
    @buffer.fill(nil)
    @read_index  = 0
    @write_index = 0
  end

  def read
    raise BufferEmptyException if @buffer[@read_index].nil?
    value = @buffer[@read_index]
    @buffer[@read_index] = nil
    @read_index = advance(@read_index)
    value
  end

  def write(value)
    raise BufferFullException if !@buffer[@write_index].nil?
    @buffer[@write_index] = value
    @write_index = advance(@write_index)
  end

  def write!(value)
    return write(value) if @buffer.any?(&:nil?)
    @buffer[@read_index] = value
    @read_index = advance(@read_index)
  end

  private

  def advance(index)
    return 0 if index == @last_index
    index + 1
  end
end

class CircularBuffer::BufferEmptyException < StandardError
  def initialize
    super("Cannot read from an empty buffer.")
  end
end

class CircularBuffer::BufferFullException < StandardError
  def initialize
    super("Cannot write to a full buffer.")
  end
end
